### PR TITLE
Sharding-jdbc supports nacos configuration

### LINKED ISSUE #25445: Sharding-jdbc supports nacos configuration
## Question

Support for nacos configuration has been achieved through SPI

## Todo

- [x] Add NacosDriverURLProvider.java files to org.apache.shardingsphere.driver.jdbc.core.driver.spi
- [x] Add the configuration of NacosDriverURLProvider in org.apache.shardingsphere.driver.jdbc.core.driver.ShardingSphereDriverURLProvider
- [x] Add nacos dependency in pom file
- [x] Add configuration examples
### Detail
NacosDriverURLProvider.java
```
package org.apache.shardingsphere.driver.jdbc.core.driver.spi;

import com.alibaba.nacos.api.NacosFactory;
import com.alibaba.nacos.api.common.Constants;
import com.alibaba.nacos.api.config.ConfigService;
import com.alibaba.nacos.api.exception.NacosException;
import com.google.common.base.Preconditions;
import lombok.SneakyThrows;
import org.apache.commons.lang3.StringUtils;
import org.apache.shardingsphere.driver.jdbc.core.driver.ShardingSphereDriverURLProvider;

import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;

public class NacosDriverURLProvider implements ShardingSphereDriverURLProvider {

    private static final String PARAM_PREFIX = "?";
    private static final String NACOS_TYPE = "nacos:";
    private static final String SHRDING_URL_PREFIX = "jdbc:shardingsphere:";

    @Override
    public boolean accept(String url) {
        return StringUtils.isNotBlank(url) && url.contains(NACOS_TYPE);
    }

    @Override
    @SneakyThrows(NacosException.class)
    public byte[] getContent(String url) {
        Properties props = getNacosConfigByUrl(url);
        ConfigService configService = NacosFactory.createConfigService(props);
        String dataId = props.getProperty(Constants.DATAID);
        String config = configService.getConfig(dataId, props.getProperty(Constants.GROUP, Constants.DEFAULT_GROUP), 500);
        Preconditions.checkArgument(config != null, "Nacos config [" + dataId + "] is Empty.");
        return config.getBytes(StandardCharsets.UTF_8);
    }

    private Properties getNacosConfigByUrl(String url) {
        String nacosConfig = StringUtils.removeStart(url, SHRDING_URL_PREFIX + NACOS_TYPE);
        Preconditions.checkArgument(nacosConfig.indexOf(PARAM_PREFIX) > 0, "Nacos param is required in ShardingSphere driver URL.");
        String dataId = StringUtils.substringBefore(nacosConfig, PARAM_PREFIX);
        //'withKeyValueSeparator(java.lang.String)' is marked unstable with @Beta
        //Map<String, String> confMap = Splitter.on("&").withKeyValueSeparator("=").split(nacosConfig);
        Map<String, String> confMap = getUrlParameters(StringUtils.substringAfter(nacosConfig, PARAM_PREFIX));
        Properties properties = new Properties();
        properties.putAll(confMap);
        properties.put(Constants.DATAID, dataId);
        return properties;
    }

    public static Map<String, String> getUrlParameters(String urlParam) {
        Map<String, String> mapRequest = new HashMap<>();
        String[] arrSplit;
        if (urlParam == null) {
            return mapRequest;
        }
        arrSplit = urlParam.split("[&]");
        for (String strSplit : arrSplit) {
            String[] arrSplitEqual;
            arrSplitEqual = strSplit.split("[=]");
            if (arrSplitEqual.length > 1) {
                mapRequest.put(arrSplitEqual[0], arrSplitEqual[1]);
            } else {
                if (!"".equals(arrSplitEqual[0])) {
                    mapRequest.put(arrSplitEqual[0], "");
                }
            }
        }
        return mapRequest;
    }

}
```
org.apache.shardingsphere.driver.jdbc.core.driver.ShardingSphereDriverURLProvider
```text
org.apache.shardingsphere.driver.jdbc.core.driver.spi.NacosDriverURLProvider
```

Add dependencies to pom files in shardingsphere
```
<nacos-client.version>1.4.2</nacos-client.version>

<dependency>
         <groupId>com.alibaba.nacos</groupId>
         <artifactId>nacos-client</artifactId>
         <version>${nacos-client.version}</version>
         <scope>provided</scope>
</dependency>
```

Add dependencies to pom files in shardingsphere-jdbc-core
```
<dependency>
            <groupId>com.alibaba.nacos</groupId>
            <artifactId>nacos-client</artifactId>
</dependency>
```
Example
```
url: jdbc:shardingsphere:nacos:sharding-config-test.yaml?serverAddr=172.16.11.45:8848&namespace=dev&group=DEFAULT_GROUP&username=nacos&password=nacos
```

### PR DESCRIPTION
**Pull request by liming0**
Fixes #25445.

Changes proposed in this pull request:
  -

---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [x] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [x] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: docs/document/content/dev-manual/data-source.cn.md [hunk 1]
```diff
@@ -118,7 +118,7 @@ ShardingSphere 驱动 URL 提供器
...
 | jdbc:shardingsphere:absolutepath:<path> | 驱动的绝对路径加载器     | [`org.apache.shardingsphere.driver.jdbc.core.driver.spi.AbsolutePathDriverURLProvider`](https://github.com/apache/shardingsphere/blob/master/jdbc/core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/driver/spi/AbsolutePathDriverURLProvider.java) |
 | jdbc:shardingsphere:apollo:<namespace>  | 驱动的 Apollo 加载器 | [`org.apache.shardingsphere.driver.jdbc.core.driver.spi.ApolloDriverURLProvider`](https://github.com/apache/shardingsphere/blob/master/jdbc/core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/driver/spi/ApolloDriverURLProvider.java)             |
-
+| jdbc:shardingsphere:nacos:<param>      | 驱动的 Nacos 加载器 | [`org.apache.shardingsphere.driver.jdbc.core.driver.spi.ApolloDriverURLProvider`](https://github.com/apache/shardingsphere/blob/master/jdbc/core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/driver/spi/NacosDriverURLProvider.java)             |
```
Comment (zhaojinchao95) [inline docs/document/content/dev-manual/data-source.cn.md:121]: Is it `NacosDriverURLProvider`?

File: jdbc/core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/driver/spi/NacosDriverURLProvider.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        String config = configService.getConfig(dataId, props.getProperty(Constants.GROUP, Constants.DEFAULT_GROUP), 500);
+        Preconditions.checkArgument(config != null, "Nacos config [" + dataId + "] is Empty.");
+        return config.getBytes(StandardCharsets.UTF_8);
+    }
+
+    private Properties getNacosConfigByUrl(final String url) {
+        String nacosConfig = StringUtils.removeStart(url, SHRDING_URL_PREFIX + NACOS_TYPE);
+        Preconditions.checkArgument(nacosConfig.indexOf(PARAM_PREFIX) > 0, "Nacos param is required in ShardingSphere driver URL.");
+        String dataId = StringUtils.substringBefore(nacosConfig, PARAM_PREFIX);
+        //'withKeyValueSeparator(java.lang.String)' is marked unstable with @Beta
```
Comment (zhaojinchao95) [inline jdbc/core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/driver/spi/NacosDriverURLProvider.java:62]: Please remove this comment.

File: jdbc/core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/driver/spi/NacosDriverURLProvider.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        Properties properties = new Properties();
+        properties.putAll(confMap);
+        properties.put(Constants.DATAID, dataId);
+        return properties;
+    }
+
+    /**
+     * Parse url parameters.
+     *
+     * @param urlParam 'aaa=111&bbb=222'.
```
Comment (zhaojinchao95) [inline jdbc/core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/driver/spi/NacosDriverURLProvider.java:74]: You can refer to other comment, remove `@author` and `@date`

File: jdbc/core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/driver/spi/NacosDriverURLProvider.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            arrSplitEqual = strSplit.split("[=]");
+            if (arrSplitEqual.length > 1) {
+                mapRequest.put(arrSplitEqual[0], arrSplitEqual[1]);
+            } else {
+                if (!"".equals(arrSplitEqual[0])) {
+                    mapRequest.put(arrSplitEqual[0], "");
+                }
+            }
+        }
+        return mapRequest;
```
Comment (zhaojinchao95) [inline jdbc/core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/driver/spi/NacosDriverURLProvider.java:97]: Change `mapRequest` to `result`.

**Comment by liming0** [comment]
image
@zhaojinchao95 Hello, all the modifications have been completed, but there is an error that I don't know how to solve. This modification does not involve any modifications to the module. Can you help me specify it

**Comment by zhaojinchao95** [comment]
> image @zhaojinchao95 Hello, all the modifications have been completed, but there is an error that I don't know how to solve. This modification does not involve any modifications to the module. Can you help me specify it

You can look this: 

image

**Comment by zhaojinchao95** [comment]
You can compile it at your local environment, The command is : `mvn clean install -Prelease -T1C -Djacoco.skip=true -Drat.skip=true -Dmaven.javadoc.skip=true -B`

**Comment by zhaojinchao95** [comment]
@liming0 Hi, Can you fix it?

**Comment by liming0** [comment]
> You can compile it at your local environment, The command is : `mvn clean install -Prelease -T1C -Djacoco.skip=true -Drat.skip=true -Dmaven.javadoc.skip=true -B`

Sorry, I encountered a problem while executing this command locally, but I haven't used this signed plugin before and don't know how to solve it
```
Failed to execute goal org.apache.maven.plugins:maven-gpg-plugin:3.0.1:sign (sign-release-artifacts) on project shardingsphere: Execution sign-release-artifacts of goal org.apache.maven.plugins:maven-gpg-plugin:3.0.1:sign failed.
```

**Comment by zhaojinchao95** [comment]
> > You can compile it at your local environment, The command is : `mvn clean install -Prelease -T1C -Djacoco.skip=true -Drat.skip=true -Dmaven.javadoc.skip=true -B`
> > @zhaojinchao95
> > Sorry, I encountered a problem while executing this command locally, but I haven't used this signed plugin before and don't know how to solve it
> 
> ```
> Failed to execute goal org.apache.maven.plugins:maven-gpg-plugin:3.0.1:sign (sign-release-artifacts) on project shardingsphere: Execution sign-release-artifacts of goal org.apache.maven.plugins:maven-gpg-plugin:3.0.1:sign failed.
> ```

I haven't met this problem. You can run `JDBCMetaDataInfoExporterTest.class `

**Comment by vanniuner** [comment]
what about this issue ? i'm waiting to use!

**Comment by boyjoy1127** [comment]
Will this issue be completed before June 15 which is the final date of version 5.4.0?

**Comment by liming0** [comment]
> Will this issue be completed before June 15 which is the final date of version 5.4.0?
I'm afraid not. The code is already completed, but there are compilation issues. Looking forward to help😂

**Comment by liming0** [comment]
> Will this issue be completed before June 15 which is the final date of version 5.4.0?

I did not succeed in the CI phase, but there is a compilation issue with this project on my computer, which seems to be related to those. G4 files. I don't know how to solve it. Can you help me?

**Pull request closed without merging**